src/shared/awsagent/roles.go

```diff
@@ -173,48 +175,185 @@ func (a *Agent) CreateRolesAnywhereProfileForRole(ctx context.Context, role type
-	logger := logrus.WithField("namespace", namespaceName).WithField("account", accountName)
-	exists, role, err := a.GetOtterizeRole(ctx, namespaceName, accountName)
+func (a *Agent) IsPolicyDocumentsIdenticalUnordered(role *types.Role, newDocument PolicyDocument) (bool, error) {
+	if role.AssumeRolePolicyDocument == nil {
+		return false, errors.New("role AssumeRolePolicyDocument is nil")
+	}
 
+	// AssumeRolePolicyDocument is url-encoded
+	assumeRolePolicyDocumentRaw, err := url.QueryUnescape(*role.AssumeRolePolicyDocument)
 	if err != nil {
-		return nil, errors.Wrap(err)
+		return false, errors.Wrap(err)
 	}
 
-	if exists {
-		logger.Debugf("found existing role, arn: %s", *role.Arn)
-		// check if it is soft deleted - if so remove soft deleted tag
-		if hasSoftDeletedTagSet(role.Tags) {
-			logger.Debug("role is tagged unused, untagging")
-			// There is no need to untag the role's policies from this context, It will happen if the policy will be created again
-			_, err := a.iamClient.UntagRole(ctx, &iam.UntagRoleInput{RoleName: role.RoleName, TagKeys: []string{softDeletedTagKey}})
-			if err != nil {
-				return nil, errors.Wrap(err)
+	var existingPolicy map[string]any
+	err = json.Unmarshal([]byte(assumeRolePolicyDocumentRaw), &existingPolicy)
+	if err != nil {
+		return false, errors.Wrap(err)
+	}
+
+	// For some reason, Action can either be a list or a string.
+	statements := existingPolicy["Statement"].([]any)
+	for i, statement := range statements {
+		statementTyped := statement.(map[string]any)
+		// Get action or actions
+		if action, ok := statementTyped["Action"]; ok {
+			switch actionType := action.(type) {
+			case string:
+				statementTyped["Action"] = []string{actionType} // Convert single string to slice
+				statements[i] = statementTyped                  // Update the statement in the slice
+			case []any:
+				// do nothing
+			default:
+				return false, errors.Errorf("unexpected type for Action: %T", actionType)
 			}
-			// Intentionally not returning here, as we want to continue and check if we need to mark as soft delete only
 		}
-		if useSoftDeleteStrategy {
-			err = a.setRoleSoftDeleteStrategyTag(ctx, role)
-			if err != nil {
-				return nil, errors.Wrap(err)
+	}
+	existingPolicy["Statement"] = statements
+
+	// Marshal the policy back to JSON
+	assumeRolePolicyDocument, err := json.Marshal(existingPolicy)
+	if err != nil {
+		return false, errors.Wrap(err)
+	}
+
+	var existingPolicyStructured PolicyDocument
+	err = json.Unmarshal(assumeRolePolicyDocument, &existingPolicyStructured)
+	if err != nil {
+		return false, errors.Wrap(err)
+	}
+
+	if existingPolicyStructured.Version != newDocument.Version {
+		return false, nil
+	}
+
+	if len(existingPolicyStructured.Statement) != len(newDocument.Statement) {
+		return false, nil
+	}
+
+	for _, newStatement := range newDocument.Statement {
+		found := false
+		for _, existingStatement := range existingPolicyStructured.Statement {
```

Comment:
move the comparison between new and existing stements to a dedicated function to make this code nicer (and avoid confusing `continue` pattern): 
```
	for _, newStatement := range newDocument.Statement {
		found := false
		for _, existingStatement := range existingPolicyStructured.Statement {
			// compareStatements returns true if stements are equal
			if compareStatements(newStatement, existingStatement) {
				found = true
				break
			}
		if !found {...} 
```

or even nicer:
```
	for _, newStatement := range newDocument.Statement {
		existingStatement, found := lo.FindBy(existingPolicyStructured.Statement, func (....) {return compareStatements(existingStatement, newStatement) }
		if !found {...} 
```